### PR TITLE
avm2: Partially implement Loader.load

### PR DESCRIPTION
**Pull request by Aaron1011**
Depends on https://github.com/ruffle-rs/ruffle/pull/7633

This PR implements the `Loader.load` method, as well as
the associated `LoaderInfo` properties and events.

We can now load in an external AVM2 SWf: it will be added
as a child of `Loader` object, and will render properly
to the screen.

Limitations:
* The only supported `URLRequest` property is `url`
* `LoaderContext` is not supported at all - we always use the default
  behavior
* Only `Loader.load` is implemented - we do not yet support unloading.
* We fire a plain 'Event' for the 'progress' event, instead of using
  the (not yet implemented) 'ProgressEvent' class

The main changes in this PR are:
* The AVM2 `Loader` class now has an associated display object,
  `LoaderDisplay`. This is basically a stub, and just renders
  its single child (if it exists).
* `LoaderStream::Stage` is renamed to `LoaderStream::NotYetLoaded`.
  This is used for both the `Stage` and an 'uninitialized'
  `Loader.contentLoaderInfo`. In both cases, certain properties throw
  errors, while others return actual values.
* The rust `Loader` manager now handles both AVM1 and AVM2 movie loads.

### CODE REVIEW

**Comment by adrian17** [comment]
> instead of using the (not yet implemented) 'ProgressEvent' class

It's implemented :)

**Comment by Aaron1011** [comment]
@adrian17 I've updated the implementation to correctly handle the "progress" event.

File: core/src/avm2/globals.rs [hunk 1]
```diff
@@ -89,13 +89,16 @@ pub struct SystemClasses<'gc> {
...
+    pub progressevent: ClassObject<'gc>,
     pub textevent: ClassObject<'gc>,
     pub errorevent: ClassObject<'gc>,
     pub ioerrorevent: ClassObject<'gc>,
     pub securityerrorevent: ClassObject<'gc>,
     pub transform: ClassObject<'gc>,
     pub colortransform: ClassObject<'gc>,
     pub matrix: ClassObject<'gc>,
+    pub illegaloperationerror: ClassObject<'gc>,
+    pub event_dispatcher: ClassObject<'gc>,
```
Comment (Herschel) [inline core/src/avm2/globals.rs:101]: Should this be `eventdispatcher` to match the naming convention?

(Separate from this PR, should we use `snake_case` the naming convention for all of these eventually?)

File: swf/src/read.rs [hunk 2]
```diff
@@ -2589,7 +2589,7 @@ pub fn read_compression_type<R: Read>(mut input: R) -> Result<Compression> {
         b"FWS" => Compression::None,
         b"CWS" => Compression::Zlib,
         b"ZWS" => Compression::Lzma,
-        _ => return Err(Error::invalid_data("Invalid SWF")),
+        _ => return Err(Error::invalid_data(format!("Invalid SWF: {:?}", signature))),
```
Comment (Herschel) [inline swf/src/read.rs:2592]: Seems like this isn't very helpful to print the signature? It would be gibberish for any non-SWF file.

**Review by Herschel — APPROVED** [review]

**Comment by Herschel** [comment]
Thank you!

File: core/src/avm2/globals.rs (hunk 1 above)
Comment (Aaron1011) [reply core/src/avm2/globals.rs:101]: I think snake case would be better in the long run, but I'm fine with changing this to `eventdispatcher` for now.

File: core/src/avm2/globals/flash/display/loader.rs [hunk 3]
```diff
@@ -1,71 +1,91 @@
...
+        if let Some(context) = args.get(1) {
+            if !matches!(context, Value::Null) {
+                log::warn!(
+                    "Loader.load: 'context' argument is not yet implemented: {:?}",
+                    context
+                );
+            }
+        }
+        let url = url_request
+            .get_property(&QName::dynamic_name("url").into(), activation)?
```
Comment (adrian17) [inline core/src/avm2/globals/flash/display/loader.rs:63]: let's use `&Multiname::public("url")` in new code, this is the more direct construction

File: core/src/avm2/object/loaderinfo_object.rs [hunk 4]
```diff
@@ -147,17 +169,52 @@ impl<'gc> TObject<'gc> for LoaderInfoObject<'gc> {
                     e
                 );
             }
+
+            let complete_evt = EventObject::bare_default_event(context, "complete");
+
+            if let Err(e) = Avm2::dispatch_event(context, complete_evt, (*self).into()) {
+                log::error!(
+                    "Encountered AVM2 error when broadcasting `complete` event: {}",
```
Comment (adrian17) [inline core/src/avm2/object/loaderinfo_object.rs:177]: side note: maybe we should push this into `dispatch_event`? As it's currently duplicated many times already, I think

File: core/src/display_object/movie_clip.rs [hunk 5]
```diff
@@ -946,6 +966,10 @@ impl<'gc> MovieClip<'gc> {
         self.0.read().static_data.total_frames
     }
 
+    pub fn avm2_class(self) -> Option<Avm2ClassObject<'gc>> {
```
Comment (adrian17) [inline core/src/display_object/movie_clip.rs:969]: is this used anywhere?

File: core/src/avm2/globals/flash/display/loader.rs (hunk 3 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/display/loader.rs]: I've adjusted this PR to use `&Multiname::public` where applicable.

**Review by adrian17 — APPROVED** [review]
LGTM, but would still prefer if Kmeisthax or Mike took a look before merging.

File: core/src/avm2/object/loaderinfo_object.rs (hunk 4 above)
Comment (Aaron1011) [reply core/src/avm2/object/loaderinfo_object.rs:177]: Yeah, I was thinking about doing that in a follow-up PR.

File: core/src/display_object/movie_clip.rs (hunk 5 above)
Comment (Aaron1011) [reply core/src/display_object/movie_clip.rs]: It's not - I've removed it.

**Pull request merged**